### ISSUE TITLE
Unable to run go_ibm_db@v0.4.2 on M1 architecture

### ISSUE THREAD

**Issue by fbasit-ibm** [ibmdb/go_ibm_db#194]
Our team makes heavy use of IBM's golang driver for DB2 (https://github.com/ibmdb/go_ibm_db). In the past, I've had no problems using the driver on my Intel macbook pro however I recently upgraded my machine to Apple Silicon and have been getting the same error consistently. Also I noticed there is only one macos clidriver available (https://public.dhe.ibm.com/ibmdl/export/pub/software/data/db2/drivers/odbc_cli/), not sure if this is meant to work with both intel/apple silicon. Thank you again, any help is appreciated.

- arch: M1 Max
- Golang: go1.18.8 darwin/arm64
- go_ibm_db v0.4.2
- IBM_DB_HOME=/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver
- CGO_CFLAGS=-I/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/include
- CGO_LDFLAGS=-L/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/lib
- DYLD_LIBRARY_PATH=:/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/lib

<details><summary><code>go env</code> Output:</summary><br><pre>
$ go test -p 1 -v -coverpkg=./internal/... -coverprofile=test/coverage/unit_test.tmp github.ibm.com/Activity-Insights/analytics-extract/internal/helpers

github.com/ibmdb/go_ibm_db/api
ld: warning: ignoring file /Users/furqan/go/pkg/mod/github.com/ibmdb/clidriver/lib/libdb2.dylib, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
Undefined symbols for architecture arm64:
  "_SQLAllocHandle", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLAllocHandle in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLAllocHandle)
  "_SQLBindCol", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLBindCol in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLBindCol)
  "_SQLBindParameter", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLBindParameter in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLBindParameter)
  "_SQLCloseCursor", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLCloseCursor in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLCloseCursor)
  "_SQLColAttribute", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLColAttribute in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLColAttribute)
  "_SQLCreateDbW", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLCreateDbW in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLCreateDbW)
  "_SQLDescribeColW", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLDescribeColW in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLDescribeColW)
  "_SQLDescribeParam", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLDescribeParam in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLDescribeParam)
  "_SQLDisconnect", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLDisconnect in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLDisconnect)
  "_SQLDriverConnectW", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLDriverConnectW in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLDriverConnectW)
  "_SQLDropDbW", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLDropDbW in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLDropDbW)
  "_SQLEndTran", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLEndTran in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLEndTran)
  "_SQLExecDirectW", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLExecDirectW in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLExecDirectW)
  "_SQLExecute", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLExecute in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLExecute)
  "_SQLFetch", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLFetch in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLFetch)
  "_SQLFreeHandle", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLFreeHandle in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLFreeHandle)
  "_SQLGetData", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLGetData in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLGetData)
  "_SQLGetDiagRecW", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLGetDiagRecW in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLGetDiagRecW)
  "_SQLMoreResults", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLMoreResults in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLMoreResults)
  "_SQLNumParams", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLNumParams in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLNumParams)
  "_SQLNumResultCols", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLNumResultCols in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLNumResultCols)
  "_SQLPrepareW", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLPrepareW in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLPrepareW)
  "_SQLRowCount", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLRowCount in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLRowCount)
  "_SQLSetConnectAttrW", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLSetConnectAttrW in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLSetConnectAttrW)
  "_SQLSetEnvAttr", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLSetEnvAttr in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLSetEnvAttr)
  "_SQLSetStmtAttrW", referenced from:
      __cgo_9b381ce23318_Cfunc_SQLSetStmtAttrW in _x003.o
     (maybe you meant: __cgo_9b381ce23318_Cfunc_SQLSetStmtAttrW)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
FAIL	github.ibm.com/Activity-Insights/analytics-extract/internal/helpers [build failed]
FAIL

</pre></details>

Steps to Reproduce:

1. Downloading the macos clidriver from here [macos64_odbc_cli.tar.gz](https://public.dhe.ibm.com/ibmdl/export/pub/software/data/db2/drivers/odbc_cli/macos64_odbc_cli.tar.gz)

**Comment by fbasit-ibm** [comment]
I also noticed this https://github.com/ibmdb/python-ibmdb/issues/712. This user was able to use the python driver with M1. I tried emulating a similar fix with golang, but I had no luck

**Comment by bimalkjha** [comment]
@fbasit-ibm You need to install x86-64 version of golang in your M1 system, then only ibm_db will work. ibm_db do not work with arm64 golang. You can replace your installed golang with this one https://go.dev/dl/go1.19.3.darwin-amd64.pkg and ibm_db should work fine. Thanks.

**Comment by fbasit-ibm** [comment]
@bimalkjha Thanks for your reply. I tried installing both golang versions amd64 and arm64 with no luck. I got the same error.

**Comment by fbasit-ibm** [comment]
Actually, using the x86-64 version now gives me a different error message:

<summary><code>go env</code> Output:</summary><br><pre>
$ go test -p 1 -v -coverpkg=./internal/... -coverprofile=test/coverage/unit_test.tmp github.ibm.com/Activity-Insights/analytics-extract/internal/helpers
dyld[14301]: Library not loaded: libdb2.dylib
  Referenced from: <431DDC87-6150-3C05-B7BD-B02F4E6FADBF> /private/var/folders/17/k5j7xtfs60b3_vdjbsv1vfdc0000gn/T/go-build3023464420/b001/helpers.test
  Reason: tried: 'libdb2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibdb2.dylib' (no such file), 'libdb2.dylib' (no such file), '/usr/local/lib/libdb2.dylib' (no such file), '/usr/lib/libdb2.dylib' (no such file, not in dyld cache), '/Users/furqan/Documents/guardium_insights/analytics-extract/internal/helpers/libdb2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/furqan/Documents/guardium_insights/analytics-extract/internal/helpers/libdb2.dylib' (no such file), '/Users/furqan/Documents/guardium_insights/analytics-extract/internal/helpers/libdb2.dylib' (no such file), '/usr/local/lib/libdb2.dylib' (no such file), '/usr/lib/libdb2.dylib' (no such file, not in dyld cache)
FAIL	github.ibm.com/Activity-Insights/analytics-extract/internal/helpers	0.859s
FAIL
</pre>

**Comment by vmathur12** [comment]
@fbasit-ibm Create a symbolic link in the directory to load the “libdb2.dylib” library.
Use the following command:
ln -s /Users/<uname>/go/src/github.com/ibmdb/clidriver/lib/libdb2.dylib libdb2.dylib
Thanks.

**Comment by fbasit-ibm** [comment]
@vmathur12 Unfortunately I still get the same error

<summary><code>go env</code> Output:</summary><br><pre>
$ ln -s /Users/furqan/go/pkg/mod/github.com/ibmdb/clidriver/lib/libdb2.dylib libdb2.dylib
$ go test -p 1 -v -coverpkg=./internal/... -coverprofile=test/coverage/unit_test.tmp github.ibm.com/Activity-Insights/analytics-extract/internal/helpers
dyld[58798]: Library not loaded: libdb2.dylib
  Referenced from: <D227B2F8-A326-3573-96B3-DF658146FF42> /private/var/folders/17/k5j7xtfs60b3_vdjbsv1vfdc0000gn/T/go-build2634122810/b001/helpers.test
  Reason: tried: 'libdb2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibdb2.dylib' (no such file), 'libdb2.dylib' (no such file), '/usr/local/lib/libdb2.dylib' (no such file), '/usr/lib/libdb2.dylib' (no such file, not in dyld cache), '/Users/furqan/Documents/guardium_insights/analytics-extract/internal/helpers/libdb2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/furqan/Documents/guardium_insights/analytics-extract/internal/helpers/libdb2.dylib' (no such file), '/Users/furqan/Documents/guardium_insights/analytics-extract/internal/helpers/libdb2.dylib' (no such file), '/usr/local/lib/libdb2.dylib' (no such file), '/usr/lib/libdb2.dylib' (no such file, not in dyld cache)
FAIL	github.ibm.com/Activity-Insights/analytics-extract/internal/helpers	0.976s
FAIL
</pre>

**Comment by fbasit-ibm** [comment]
Is the new go_ibm_db driver only compatible with golang version 1.19+? Our code uses 1.18 to compile so it's not completely possible to switch to 1.19 at the moment for me

**Comment by vmathur12** [comment]
@fbasit-ibm  Could you please share following output of the commands. 
go env
ls -l ... clidriver
ls -l .. clidriver/lib
ls -l .. testdata (test folder)
for env in IBM_DB_HOME=$IBM_DB_HOME CGO_CFLAGS=$CGO_CFLAGS CGO_LDFLAGS=$CGO_LDFLAGS DYDLD_LIBRARY_PATH$DYLD_LIBRARY_PATH; do echo $env; done

I am not seeing the error in my environment.

Output:
go version go1.18.3 darwin/amd64
go_ibm_db v0.4.2

IBM_DB_HOME=/Users/uname/go/pkg/mod/github.com/ibmdb/clidriver
CGO_CFLAGS=-I/Users/uname/go/pkg/mod/github.com/ibmdb/clidriver/include
CGO_LDFLAGS=-L/Users/uname/go/pkg/mod/github.com/ibmdb/clidriver/lib
DYDLD_LIBRARY_PATH:/Users/uname/go/pkg/mod/github.com/ibmdb/clidriver/lib

$ ls -l
-rwxr-xr-x     134 Nov 25 08:46 Arraybool_test.go
-rwxr-xr-x     129 Nov 25 08:46 Arraychar_test.go
-rwxr-xr-x     132 Nov 25 08:46 Arrayfloat_test.go
-rw-r--r--      65 Nov 25 08:45 go.mod
-rw-r--r--     171 Nov 25 08:45 go.sum
lrwxr-xr-x      69 Nov 25 08:50 libdb2.dylib -> /Users/uname/go/pkg/mod/github.com/ibmdb/clidriver/lib/libdb2.dylib
-rwxr-xr-x   28803 Nov 25 08:45 main.go

$ go test -v
=== RUN   TestBoolArray
--- PASS: TestBoolArray (2.63s)
=== RUN   TestCharArray
--- PASS: TestCharArray (2.56s)
=== RUN   TestFloatArray
--- PASS: TestFloatArray (2.86s)
PASS
ok      test1   8.619s

If I rename libdb2.dylib to libdb2.dylib_1, getting the similar error message
$ mv libdb2.dylib  libdb2.dylib_1
ls -l
total 104
-rwxr-xr-x      134 Nov 25 08:46 Arraybool_test.go
-rwxr-xr-x      129 Nov 25 08:46 Arraychar_test.go
-rwxr-xr-x      132 Nov 25 08:46 Arrayfloat_test.go
-rw-r--r--       65 Nov 25 08:45 go.mod
-rw-r--r--      171 Nov 25 08:45 go.sum
lrwxr-xr-x       69 Nov 25 08:50 libdb2.dylib_1 -> /Users/uname/go/pkg/mod/github.com/ibmdb/clidriver/lib/libdb2.dylib
-rwxr-xr-x    28803 Nov 25 08:45 main.go

$ go test -v
dyld[7535]: Library not loaded: libdb2.dylib
  Referenced from: /private/var/folders/3q/71_k1z0n3rbb345g5bh47ymmfww5_8/T/go-build4257238351/b001/test1.test
  Reason: tried: 'libdb2.dylib' (no such file), '/usr/local/lib/libdb2.dylib' (no such file), '/usr/lib/libdb2.dylib' (no such file), '/Users/vmathur/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/testdata1/libdb2.dylib' (no such file), '/usr/local/lib/libdb2.dylib' (no such file), '/usr/lib/libdb2.dylib' (no such file)
signal: abort trap
FAIL    test1   0.271s

**Comment by fbasit-ibm** [comment]
@vmathur12 Here are the values you requested.

```
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/furqan/Library/Caches/go-build"
GOENV="/Users/furqan/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/furqan/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/furqan/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="off"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.18.8"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-I/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/include"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-L/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/lib"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/17/k5j7xtfs60b3_vdjbsv1vfdc0000gn/T/go-build2257873171=/tmp/go-build -gno-record-gcc-switches -fno-common

```
```
IBM_DB_HOME=/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver
CGO_CFLAGS=-I/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/include
CGO_LDFLAGS=-L/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/lib
DYDLD_LIBRARY_PATH:/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/lib

```
```

$  ls -l .. clidriver
total 0
drwxr-xr-x   9 furqan  staff  288 10 Mar  2022 bin
drwxr-xr-x   8 furqan  staff  256 10 Mar  2022 bnd
drwxr-xr-x   6 furqan  staff  192  7 Mar  2022 cfg
drwxr-xr-x   3 furqan  staff   96 31 May 10:37 cfgcache
drwxr-xr-x   4 furqan  staff  128 29 May 13:43 conv
drwxr-xr-x   3 furqan  staff   96 29 May 13:43 db2dump
drwxr-xr-x  14 furqan  staff  448  7 Mar  2022 include
drwxr-xr-x   9 furqan  staff  288  9 Nov 02:04 lib
drwxr-xr-x   5 furqan  staff  160  7 Mar  2022 license
drwxr-xr-x   3 furqan  staff   96 29 May 13:43 msg
drwxr-xr-x   3 furqan  staff   96 29 May 13:43 security64

```
```
$ ls -l .. clidriver/lib
total 75736
drwxr-xr-x  17 furqan  staff       544 29 May 13:43 icc
-r-xr-xr-x   1 furqan  staff   3372520  7 Mar  2022 libDB2xml4c.58.0.dylib
lrwxr-xr-x   1 furqan  staff        22 29 May 13:43 libDB2xml4c.58.dylib -> libDB2xml4c.58.0.dylib
lrwxr-xr-x   1 furqan  staff        22 29 May 13:43 libDB2xml4c.dylib -> libDB2xml4c.58.0.dylib
-r-xr-xr-x   1 furqan  staff  31909864 31 May 10:06 libdb2.dylib
-r-xr-xr-x   1 furqan  staff     63536  7 Mar  2022 libdb2clixml4c.dylib
-r-xr-xr-x   1 furqan  staff   3422608 31 May 10:05 libstdc++.6.dylib
```

```
$ ls -l ... testdata
ls -l
total 344
-rwxr-xr-x  1 furqan  staff    134 24 Nov 12:20 Arraybool_test.go
-rwxr-xr-x  1 furqan  staff    129 24 Nov 12:20 Arraychar_test.go
-rwxr-xr-x  1 furqan  staff    132 24 Nov 12:20 Arrayfloat_test.go
-rwxr-xr-x  1 furqan  staff    126 24 Nov 12:20 Arrayint_test.go
-rwxr-xr-x  1 furqan  staff    140 24 Nov 12:20 Arraystring_test.go
-rwxr-xr-x  1 furqan  staff    146 24 Nov 12:20 Arraytimestamp_test.go
-rwxr-xr-x  1 furqan  staff    196 24 Nov 12:20 ConnectionPoolWithTimeout_test.go
-rwxr-xr-x  1 furqan  staff    161 24 Nov 12:20 ConnectionPool_test.go
-rwxr-xr-x  1 furqan  staff    139 24 Nov 12:20 CreateDB_test.go
-rwxr-xr-x  1 furqan  staff    135 24 Nov 12:20 DropDB_test.go
-rwxr-xr-x  1 furqan  staff    132 24 Nov 12:20 ExecDirect_test.go
-rwxr-xr-x  1 furqan  staff    141 24 Nov 12:20 a_create_test.go
-rwxr-xr-x  1 furqan  staff    131 24 Nov 12:20 begin_test.go
-rwxr-xr-x  1 furqan  staff    126 24 Nov 12:20 close_test.go
-rwxr-xr-x  1 furqan  staff    132 24 Nov 12:20 columns_test.go
-rwxr-xr-x  1 furqan  staff    126 24 Nov 12:20 commit_test.go
-rwxr-xr-x  1 furqan  staff    151 24 Nov 12:20 `conn_test.go`
-rwxr-xr-x  1 furqan  staff    126 24 Nov 12:20 drop_test.go
-rwxr-xr-x  1 furqan  staff    158 24 Nov 12:20 execContext_test.go
-rwxr-xr-x  1 furqan  staff    124 24 Nov 12:20 insert_test.go
-rwxr-xr-x  1 furqan  staff  28371 24 Nov 12:20 main.go
-rwxr-xr-x  1 furqan  staff    124 24 Nov 12:20 next_test.go
-rwxr-xr-x  1 furqan  staff    146 24 Nov 12:20 nullbool_test.go
-rwxr-xr-x  1 furqan  staff    161 24 Nov 12:20 nullcharacter_test.go
-rwxr-xr-x  1 furqan  staff    149 24 Nov 12:20 nullfloat_test.go
-rwxr-xr-x  1 furqan  staff    155 24 Nov 12:20 nullinteger_test.go
-rwxr-xr-x  1 furqan  staff    152 24 Nov 12:20 nullstring_test.go
-rwxr-xr-x  1 furqan  staff    146 24 Nov 12:20 nulltime_test.go
-rwxr-xr-x  1 furqan  staff    150 24 Nov 12:20 poolopen_test.go
-rwxr-xr-x  1 furqan  staff    155 24 Nov 12:20 prepareContext_test.go
-rwxr-xr-x  1 furqan  staff    131 24 Nov 12:20 prepare_test.go
-rwxr-xr-x  1 furqan  staff    144 24 Nov 12:20 queryContext_test.go
-rwxr-xr-x  1 furqan  staff    125 24 Nov 12:20 query_test.go
-rwxr-xr-x  1 furqan  staff    132 24 Nov 12:20 queryrow_test.go
-rwxr-xr-x  1 furqan  staff    124 24 Nov 12:20 scan_test.go
-rwxr-xr-x  1 furqan  staff    158 24 Nov 12:20 spInOut_test.go
-rwxr-xr-x  1 furqan  staff    148 24 Nov 12:20 sp_test.go
```

**Comment by bimalkjha** [comment]
@fbasit-ibm In your shared output, we can see the setting for DYLD_LIBRARY_PATH is not correct.
`DYDLD_LIBRARY_PATH:/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/lib`
It should have = instead of : 
Execute command `export DYLD_LIBRARY_PATH=/Users/uname/go/pkg/mod/github.com/ibmdb/clidriver/lib` and then run go test file.
Also, make sure you have gcc installed in your system. For M1 Chip system, gcc is required. You can refer last comment in https://github.com/ibmdb/python-ibmdb/issues/797 for details about requirement to install x86-64 version of gcc and golang in the system to use with ibm_db. Thanks.

**Comment by fbasit-ibm** [comment]
Firstly, thank you @bimalkjha and @vmathur12 for your prompt assistance.

These are my findings...

## Installing gcc-8 on M1 architecture

**Warning**: If you use the ARM version of homebrew *(as recommended for M1 architectures)* you will get the following error message:
```
$ brew install gcc-8
Error: Cannot install in Homebrew on ARM processor in Intel default prefix (/usr/local)!
Please create a new installation in /opt/homebrew using one of the
"Alternative Installs" from:
  https://docs.brew.sh/Installation
You can migrate your previously installed formula list with:
  brew bundle dump
```

1. As instructed, I installed `gcc-8` using homebrew *(note: the x86_64 version of homebrew is needed for this, not the recommended ARM based homebrew)*. The clearest instructions on how to install and use the x86_64 version of homebrew is by following this users steps: https://github.com/crystal-lang/crystal/issues/10237#issuecomment-876634478.

2. I installed gcc-8 using the x86_64 homebrew:

```
ibrew install gcc@8
```

Once the installation succeeds, I had to create a `symbolic link` to override my native ARM based gcc so my system can reference gcc-8:

```
sudo ln -s /usr/local/bin/gcc-8 /usr/local/bin/gcc
```

## Installing Golang and go_ibm_db

1. I installed the x86_64 version of Golang v1.18.8 on my M1 Max machine. (https://go.dev/dl/go1.18.8.darwin-amd64.pkg).

2. I installed go_ibm_db using:
```
$ go install github.com/ibmdb/go_ibm_db
```

3. I ran the `setup.go` file to install the clidriver:
```
$ cd /Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/installer/
$ go run setup.go
```
This installed the `https://public.dhe.ibm.com/ibmdl/export/pub/software/data/db2/drivers/odbc_cli/macos64_odbc_cli.tar.gz` driver into my local machine.


4. I exported these variables into my .zsh environment:
```
IBM_DB_HOME=/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver
CGO_CFLAGS=-I/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/include
CGO_LDFLAGS=-L/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/lib
DYDLD_LIBRARY_PATH=/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/lib:/usr/local/lib/gcc/8
```

5. Once these steps were done, as a developer I'd like to think my job is done. However now, when running my go code I still get the following error:
```
$ go run main.go

dyld[58798]: Library not loaded: libdb2.dylib

Referenced from:  /private/var/folders/17/k5j7xtfs60b3_vdjbsv1vfdc0000gn/T/go-build2634122810/b001/helpers.test

Reason: tried: 'libdb2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibdb2.dylib' (no such file), 'libdb2.dylib' (no such file), '/usr/local/lib/libdb2.dylib' (no such file), '/usr/lib/libdb2.dylib' (no such file, not in dyld cache), '/Users/furqan/Documents/guardium_insights/analytics-extract/internal/helpers/libdb2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/Users/furqan/Documents/guardium_insights/analytics-extract/internal/helpers/libdb2.dylib' (no such file), '/Users/furqan/Documents/guardium_insights/analytics-extract/internal/helpers/libdb2.dylib' (no such file), '/usr/local/lib/libdb2.dylib' (no such file), '/usr/lib/libdb2.dylib' (no such file, not in dyld cache)

FAIL	github.ibm.com/Activity-Insights/analytics-extract/internal/helpers	0.976s

FAIL
```

For which the *work-around* I found was to copy the `libdb2.dylib` file from the clidriver directory into your local lib folder, as below:
```
$ cp /Users/furqan/go/pkg/mod/github.com/ibmdb/clidriver/lib/libdb2.dylib /usr/local/lib
```

Now once the library is copied over, the db2 driver doesn't throw any compilation errors. Great. However now, when using the golang the driver to connect to my DB2 instance _(which is in a healthy state)_, I get the following error:
```
$ go run ./cmd/main.go
SQLDriverConnect: {HY000} [IBM][CLI Driver] SQL10007N Message "0" could not be retrieved.  Reason code: "3".
```

The code I used for connection is sampled from this users post here https://github.com/ibmdb/go_ibm_db/issues/23#issuecomment-705381266.

# Thoughts
Now this leads me to think ... is this whole situation due to a mismatch between the go_ibm_db requiring the x86_64 version of golang versus the *(recommended)* ARM version? I'm very puzzled why this simple operation doesn't work.

Having to do so many manual steps to get the go_ibm_db driver to work on M1 architecture is a *huge* inconvenience. Is there any plan on creating a clidriver to support the M1 architecture natively? If possible, perhaps we can discuss this over a web call.

**Comment by bimalkjha** [comment]
@fbasit-ibm I see a typo in above step 4 as below:
```
DYDLD_LIBRARY_PATH=/Users/furqan/go/pkg/mod/github.com/ibmdb/go_ibm_db@v0.4.2/../clidriver/lib:/usr/local/lib/gcc/8
```
There is an extra D in above env var. I would suggest to run below commands from terminal and try:
```
export DYLD_LIBRARY_PATH=/Users/furqan/go/pkg/mod/github.com/ibmdb/clidriver/lib:/usr/local/lib/gcc/8
export PATH=/Users/furqan/go/pkg/mod/github.com/ibmdb/clidriver/bin:/Users/furqan/go/pkg/mod/github.com/ibmdb/clidriver/lib:/usr/local/lib/gcc/8:$PATH
go run ./cmd/main.go
```
If we set DYLD_LIBRARY_PATH correct, then copying libdb2.dylib to /usr/local/lib should be optional. If you still get SQL10007N  error, then remove /usr/local/lib/libdb2.dylib and try. If still error persists, then share output of below commands:
```
db2level
db2cli validate
```
Thanks.

**Comment by fbasit-ibm** [comment]
@bimalkjha Sorry, that typo was a mistake in the github comment. My environment did in fact have DYLD set correctly.
```
export DYLD_LIBRARY_PATH=/Users/furqan/go/pkg/mod/github.com/ibmdb/clidriver/lib:/usr/local/lib/gcc/8
```

The error still persists. Here are the outputs:
```
$ db2level
DB21085I  This instance or install (instance name, where applicable:
"db2inst1") uses "64" bits and DB2 code release "SQL11057" with level
identifier "0608010F".
Informational tokens are "DB2 v11.5.7.0", "special_13745",
"DYN2201131414AMD64_13745", and Fix Pack "0".
Product is installed at "/opt/ibm/db2/V11.5.0.0".

$ db2cli validate
===============================================================================
Client information for the current copy:
===============================================================================

Client Package Type       : IBM DB2 Connect Server
Client Version (level/bit): DB2 v11.5.7.0 (special_13745/64-bit)
Client Platform           : Linux/X8664
Install/Instance Path     : /opt/ibm/db2/V11.5.0.0
DB2DSDRIVER_CFG_PATH value: <not-set>
db2dsdriver.cfg Path      : /mnt/blumeta0/home/db2inst1/sqllib/cfg/db2dsdriver.cfg
DB2CLIINIPATH value       : <not-set>
db2cli.ini Path           : /mnt/blumeta0/home/db2inst1/sqllib/cfg/db2cli.ini
db2diag.log Path          : /mnt/blumeta0/db2/log/NODE0000/db2diag.log

===============================================================================
db2dsdriver.cfg schema validation for the entire file:
===============================================================================

Note: The validation utility could not find the configuration file
db2dsdriver.cfg. The file is searched at
"/mnt/blumeta0/home/db2inst1/sqllib/cfg/db2dsdriver.cfg".

===============================================================================
The validation is completed.
===============================================================================
```

**Comment by fbasit-ibm** [comment]
@bimalkjha @vmathur12 

## 01/12/22 update

I learned that GO can't read the DYLD_LIBRARY_PATH because variables LD_LIBRARY_PATH / DYLD_LIBRARY_PATH are not passed to the environment of a child process on macOS if System Integrity Protect (SIP) is enabled. This is the root of the library not found problem and also the root of the connection failure problem.

After turning off SIP, everything works as expected. In the test M1 environments that you have, I'd make a safe bet that the SIP is also disabled, hence why it always worked for you. Can you test go_ibm_db with SIP enabled on an M1 machine? I do not like the idea of having to disable SIP just for this driver to work.

**Comment by bimalkjha** [comment]
@fbasit-ibm We have released `go_ibm_db@0.5.2` with native installation support on M1 Chip system, i.e. you need not to install x64 version of go and gcc on M1 chip system to use go_ibm_db driver. We have added a fix for above SIP related issue too in the latest release.
Could you please clean up your system and remove x64 related gcc and go and other package and try fresh installation using latest driver? It should work seamlessly. Please refer the latest INSTALL.md instructions about installation on MacOS ARM64 system. Please try and update us. Thanks.
